### PR TITLE
Refactored Decoder and related classes

### PR DESCRIPTION
**Pull request by JianyuZhan**
### Description

This patch series aims at refactoring the colossal `Decoder`  class. Currently it mingles **Standard RNN** model,  **Input Feed + RNN** model, and **Transformer** model,  resulting lengthy and complicated code.

This patch series

1. Split this giant class into **StdRNNDecoder**, **InputFeedRNNDecoder**, and **TransformerDecoder**,  and replace the old `Decoder` class with a dispatcher function `make_decoder()`.
2. Shrink the `*Decoder.forward()` interfaces.

### Patches list

This series has **9** patches. All changes are splited incrementally into small patch for easy review.
Below are brief introduction, please refer to commit log of each patch for patch details.

**Belwo 3 patches are preparing patches:**
**1**. c2e76fd: test/pull_request_chk.sh: fix error
**2**.  b47d8d5: Modify TransformerDecoder constuctor for accepting options instead of `opt`
**3**.  f9b5b23 Modify Decoder constuctor for accepting options instead of `opt`

**Below 4 patches split the `Decoder` class:**
**4**. 70df07a: Skeleton patch, create the three new classes thunks and the dispatcher function. 
**5**. d31f125: Factor out  `StdRNNDecoder` and `InputFeedRNNDecoder` classes. These two
share quite a lot of code, so they inherit from a `RNNDecoderBase`  base class, in which we implement the main logics, but leave subclass-specific function to be  implemented by these classes.
**6**. 636976a: Factor out `TransformerDecoder` class.
**7**. d496a2d: Replace the old `Decoder` class with the `make_decoder()` dispatcher function, and 
                    remove the old `Decoder` class.

**Below 2 patches shrink the `forward()` interface.**
**8**. 825e323: Based on above changes.  `StdRNNDecoder` and `InputFeedRNNDecoder` naturally 
don't need the `src` argument, so easily remove it. For `TransformerDecoder` , `src` is actually used for calculating masks, and this is the same for same batch. So we pass it via `TransformerDecoderState`  when we call `init_decoder_state()` for each batch.
**9**. eedcc15: Further optimization on 7.  in `TransformerDecoder.forward(...src_words, tgt_words)`,  we repeatedly calculate the same `attn_mask` and `pad_mask` using the same `src_words` and `tgt_words` for every layer of the `TransformerDecoder`, which is a waste of time. So just calculate them and directly pass them via `TransformerDecoder.forward()`. While at it, we also modify the `TransformerEncoder`. A side-effect of this change is that we can get rid of the weird `padding_idx` in
`TransformerEncoder` and `TransformerDecoder` constructors.

### CODE REVIEW

**Comment by srush** [comment]
Rebase this with some name changes.

**Comment by JianyuZhan** [comment]
Yes, I will resend this after fixing rebase conflict.

**Comment by srush** [comment]
Also see https://github.com/OpenNMT/OpenNMT-py/pull/230 for a 4th new arch that might influence your design! Conv2Conv. Very cool.

**Comment by JianyuZhan** [comment]
I've seen it ! The code is quite independent, so I think it would make sense to create a standalone class, called `ConvDecoder` or something, just  like what TransformerDecoder is like.

Besides, now Encoder is also complecated, and also for symmetry purpose, I plan to to also split it. What do think about this?

Thanks.

**Comment by JianyuZhan** [comment]
**V1 -> V2:** 

- fixed rebase conflict.
- fixed test/pull_request_chk.sh,  due to the relocation: `test/test_model[2].pt` => `test/test_model[2].pt`

The patches structure is still the same with V1, so I just correct the commit ID in the original message.

File: onmt/Translator.py [hunk 1]
```diff
@@ -93,7 +94,8 @@ def translateBatch(self, batch, dataset):
         _, src_lengths = batch.src
         src = onmt.IO.make_features(batch, 'src')
         encStates, context = self.model.encoder(src, src_lengths)
-        decStates = self.model.init_decoder_state(context, encStates)
+        decStates = self.model.init_decoder_state(
+                src[:, :, 0].transpose(0, 1), context, encStates)
```
Comment (srush) [inline onmt/Translator.py]: This transpose is ugly, and is my fault. I implemented Transformer with batch_first indexing. 

Can you hide it in `init_decoder_state` of transformer? Just pass in `src`, and do the rest there until I fix it.

File: onmt/Translator.py [hunk 2]
```diff
@@ -64,14 +64,15 @@ def _runTarget(self, batch, data):
...
 
         #  (2) if a target is specified, compute the 'goldScore'
         #  (i.e. log likelihood) of the target under the model
         tt = torch.cuda if self.opt.cuda else torch
         goldScores = tt.FloatTensor(batch.batch_size).fill_(0)
         decOut, decStates, attn = self.model.decoder(
-            tgt_in, src, context, decStates)
+            tgt_in, context, decStates)
 
         # print(decOut.size(), batch.tgt[1:].data.size())
```
Comment (srush) [inline onmt/Translator.py]: Remove these

File: onmt/modules/Transformer.py [hunk 3]
```diff
@@ -55,60 +35,45 @@ def __init__(self, hidden_size, dropout, padding_idx,
...
         return out
 
 
 class TransformerDecoder(nn.Module):
     """
-    The Transformer Decoder from AIAYN
+    The Transformer Decoder from paper "Attetion is all you need".
     """
-    def __init__(self, hidden_size, opt, pad,
+    def __init__(self, hidden_size, dropout,
```
Comment (srush) [inline onmt/modules/Transformer.py]: Can you update these argument names. 

hidden_size -> size
d_inner -> hidden_size 
n_head -> head_count

Also can you add an assert that  `size` be divisible by `head_count` 

Thanks.

File: onmt/modules/Transformer.py [hunk 4]
```diff
@@ -55,60 +35,45 @@ def __init__(self, hidden_size, dropout, padding_idx,
...
-            self.mask = self.mask.cuda()
-        self.pad = pad
-
-    def forward(self, input, context, src_words, tgt_words):
+                                                    dropout)
+        self.dropout = dropout
+        self.mask = self._get_attn_subsequent_mask(5000)
+        # Register self.mask as a parameter of TransformerDecoder, so
+        # it gets TransformerDecoder's cuda behavior automatically.
+        self.register_parameter('mask',
```
Comment (srush) [inline onmt/modules/Transformer.py]: I think this should be a `buffer` not a `parameter`. See what was done in the new PR for self.pe in embeddings.

File: onmt/modules/Transformer.py [hunk 5]
```diff
@@ -55,60 +35,45 @@ def __init__(self, hidden_size, dropout, padding_idx,
...
-        self.dropout = opt.dropout
-        self.mask = get_attn_subsequent_mask(5000)
-        if len(opt.gpuid) > 0:
-            self.mask = self.mask.cuda()
-        self.pad = pad
-
-    def forward(self, input, context, src_words, tgt_words):
+                                                    dropout)
+        self.dropout = dropout
+        self.mask = self._get_attn_subsequent_mask(5000)
```
Comment (srush) [inline onmt/modules/Transformer.py]: Let's add a global `MAX_LEN` constant so we don't have to put 5000 everywhere.

File: onmt/modules/Transformer.py [hunk 6]
```diff
@@ -55,60 +35,45 @@ def __init__(self, hidden_size, dropout, padding_idx,
...
+        # Register self.mask as a parameter of TransformerDecoder, so
+        # it gets TransformerDecoder's cuda behavior automatically.
+        self.register_parameter('mask',
+                                nn.Parameter(self.mask, requires_grad=False))
+
+    def forward(self, input, context, attn_mask, pad_mask):
```
Comment (srush) [inline onmt/modules/Transformer.py]: These argument names are a bit off. Again my fault. 

attn_mask -> tgt_pad_mask
pad_mast -> src_pad_mask 

If you could add `aeq`'s for them that would be nice as well. Maybe you can make the `aeq` function look a bit cleaner.

File: onmt/Translator.py (hunk 1 above)
Comment (JianyuZhan) [reply onmt/Translator.py]: Okay, will do.

File: onmt/modules/Transformer.py (hunk 3 above)
Comment (JianyuZhan) [reply onmt/modules/Transformer.py]: Sure.

File: onmt/modules/Transformer.py (hunk 4 above)
Comment (JianyuZhan) [reply onmt/modules/Transformer.py]: Okay.

File: onmt/modules/Transformer.py (hunk 6 above)
Comment (JianyuZhan) [reply onmt/modules/Transformer.py]: Okay.

**Comment by JianyuZhan** [comment]
**v2->v3**

- fixed ugly `src[:, :, 0].transpose(0, 1)`,  just pass in the `src`.
- renamed argument names in `Transformer.py`  as suggested, also rename the related arguments in
`MultiHeadedAttention` and `PositionwiseFeedForward`, and add docstr to all these classes to clarify the meaning of each argument. 
- replaced `5000` with `MAX_LEN`
- used `register_buffer` for `TransformerDecoder.mask` instead of `register_parameter`.
- refactored `aeq` to explicitly accept two arguments: `base` and `*rest`,  and add more detail message
in condition of assertion failure.

The patches structure is still the same with V1/2, so I just correct the commit ID in the original message
